# NS5001

<table>
<tr>
  <td>CheckId</td>
  <td>NS5001</td>
</tr>
<tr>
  <td>Category</td>
  <td>Usage</td>
</tr>
</table>

## Cause

Awaiting callback within When method

## Rule description

A violation of this rule occurs when callback in
- `When()`
- `WhenForAnyArgs()`

is awaited

## How to fix violations

To fix a violation of this rule, use Func<T, Task> overload when using `When` or `WhenForAnyArgs`

For example:

````c#
// Incorrect:
substitute.When(async callInfo => await callInfo.Bar());

// Correct:
substitute.When(callInfo => callInfo.Bar());
````

## How to suppress violations

This warning can be suppressed by disabling the warning in the **ruleset** file for the project.
The warning can also be suppressed programmatically for an assembly:
````c#
[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "NS5001:Awaited When call.", Justification = "Reviewed")]
````

Or for a specific code block:
````c#
#pragma warning disable NS5001 // Received check.
// the code which produces warning
#pragma warning restore NS5001 // Received check.
