# NS5001

<table>
<tr>
  <td>CheckId</td>
  <td>NS5001</td>
</tr>
<tr>
  <td>Category</td>
  <td>Usage</td>
</tr>
</table>

## Cause

Argument matcher used without call to .Returns or .Received

## Rule description

Argument matchers should only be used when setting return values or checking received calls. Using `Arg.Is` or `Arg.Any` without a call to `.Returns` or `Received()` can cause your tests to behave in unexpected ways.

## How to fix violations

To fix a violation of this rule, call method or access member after received checks.

For example:

````c#
// Incorrect:
sub.Received();

// Correct:
sub.Received().SomeMethodCall();
````

## How to suppress violations

This warning can be suppressed by disabling the warning in the **ruleset** file for the project.
The warning can also be suppressed programmatically for an assembly:
````c#
[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "NS5001:Received check.", Justification = "Reviewed")]
````

Or for a specific code block:
````c#
#pragma warning disable NS5001 // Received check.
// the code which produces warning
#pragma warning restore NS5001 // Received check.
